### ISSUE TITLE
Memory leak?

### ISSUE THREAD

**Issue by IntinteDAO** [ElementsProject/elements#953]
Debian Testing (updated today)
The newest version of elements - 0.18.1.9

I am not an expert on C++, but I think Elementsd has a problem, and he is still taking care of SWAP and RAM.

Here we have data from the synchronised Liquid node. I have set up a 2GB ZRAM and Liquid has taken virtually the whole for itself. Below the java (RSK node) and Bitcoin, which at Elements are really huge!

It seems to me that Elements shouldn't take so many resources for itself.

```
elementsd 1787508 kB
java 143848 kB
bitcoind 140728 kB
unattended-upgr 6124 kB
tor 4748 kB
fail2ban-server 3464 kB
```

```
for file in /proc/*/status ; do awk '/VmSwap|Name/{printf $2 " " $3}END{ print ""}' $file; done | sort -k 2 -n -r | less
```

**Comment by apoelstra** [comment]
Can you describe on a high level what you're doing with your node? Lots of RPC calls? Creating new transactions?

There are definitely memory leaks in the RPC interface, which I fixed in #935 (which includes, among other things, much more agressive CI tools). But my impression was that they were pretty small and unlikely to add up for most usecases.

Elementsd also uses a fair bit more resident memory than Bitcoind -- about 100Mb extra by default, if I remember right. But that wouldn't explain the numbers you're seeing!

If this is an urgent problem for you I can backport my fixes, since #935 will be under review for a few more weeks at least.

**Comment by IntinteDAO** [comment]
I simply enabled the elements node (for synchronization) with transaction verification using RPC Bitcoin Core. I just synchronized the node myself today.

I need to build a master branch of "elements"?

**Comment by apoelstra** [comment]
Which version of Elements are you running? This is a surprising amount of memory usage after less than a day.

**Comment by apoelstra** [comment]
You shouldn't need to compile a new version of Elements, but if you can run the binary with `daemon=0` inside [valgrind --tool=massif](https://www.valgrind.org/docs/manual/ms-manual.html) that would be super helpful.

**Comment by IntinteDAO** [comment]
> Elements Core RPC client version elements-0.18.1.9

https://pastebin.com/pf0sffgq

It is very strange to me that he has started downloading blocks because he has been synchronised and running as a daemon for a few hours now. Maybe there is a bug in this function?

Few minutes (~5) of Elementsd work:
https://pastebin.com/sUguejua

**Comment by IntinteDAO** [comment]
Longer time:

[massif.out.11199.txt](https://github.com/ElementsProject/elements/files/5740733/massif.out.11199.txt)

**Comment by apoelstra** [comment]
Ah! So, Elements has a very weird initial block download pattern because its block headers are significantly more expensive to validate than bitcoin blocks (11 sig checks vs a couple hashes, a roughly 1000x difference). When it is initially syncing it downloads and validates all the headers, and shows no progress even in the logs (since it is a fork of Bitcoin for which this is a super fast process). Then it finishes, starts downloading blocks, and you suddenly get a slew of `UpdateTip` log messages, hours after the sync actually started.

So that explains the weird "why is it all of a sudden downloading blocks" behavior, but the memory usage is still a problem.

**Comment by apoelstra** [comment]
My guess, by the way, is that this isn't actually a memory leak, but us just incorrectly enforcing memory limits (allocating extra Elements-specific things and then not scaling our limits down correspondingly).

**Comment by stevenroose** [comment]
To reduce some of the block header overhead, it might not be a bad idea to
bring back checkpoints for Liquid. It's a federated chain anyway.

Or we could make it a bit more advanced. Since a trusted signature on a
block automatically approves all earlier blocks, you'd only have to verify
the last signature. And with dynafed, you'd only have to verify the blocks
where the consensus parameters change. I.e. the last block of each
consensus set.

**Comment by apoelstra** [comment]
I agree, I think during IBD we should (by default) validate only every 2000th block header, and dynafed transition blocks, or something like that.

In principle we could get away without validating any blocks except the last (and dynafed transitions), but it's useful for checksumming/sanity checking purposes to still validate a lot of them.

**Comment by IntinteDAO** [comment]
If you have any questions, please let me know, because I don't want to whisk elementsd and turn it on every 10 minutes...

I have set the SWAP on a 4GB frame, but that does not change much.

**Comment by IntinteDAO** [comment]
I reinstalled the system, gave 8GB Swap and no longer clogs up the ram at all.

**Comment by apoelstra** [comment]
After doing some investigation I think we do actually have an unreasonable memory spike during startup, which on my system hits 5-6Gb. The reason is than when loading the block index from disk, we de/serialize entire block headers. See [this loop](https://github.com/ElementsProject/elements/blob/master/src/txdb.cpp#L296-L311).

On Bitcoin these are 80 bytes every ten minutes, and the loop is written with that in mind (so e.g. all 700k bitcoin block headers are a bit over 50MB) but on Elements our blocks are 15 times the size and ten times the frequency. We could probably improve this.

Elements also uses more memory than Bitcoin while sitting idle, because its transactions are much larger (even ignoring the extra proofs, which are witnessdata) thanks to CT. Improving this would be harder and less valuable IMO.

**Comment by s373nZ** [comment]
I'm seeing this on `v23.3.0`:

<img width="1904" height="452" alt="Image" src="https://github.com/user-attachments/assets/8abc0454-1339-4645-ae57-e50ea6f96012" />

Recently, my node has been climbing in memory usage to the point of a crash, and Elements is one of the major suspects. Not doing anything super crazy with it. It should be hooked up to `peerswap`, yet there's not really any swapping activity.

**Comment by jsarenik** [comment]
@s373nZ wrote:
> I'm seeing this on `v23.3.0`:

I find the amount of PIDs strange. My `elementsd` keeps running as one and only process.

**Comment by s373nZ** [comment]
> I find the amount of PIDs strange. My `elementsd` keeps running as one and only process.

Thanks for the feedback. This was because I had `htop` showing userland process threads in the screenshot. There is actually only one process:

<img width="1915" height="234" alt="Image" src="https://github.com/user-attachments/assets/4136df24-99be-44f2-9210-2bae8ac18d0b" />

**Comment by a1sats** [comment]
I tested both v23.3.1 and v23.2.5, and in both cases I ran into an Out-Of-Memory error during IBD on AWS EC2 r4g.large instances. Elementsd would crash after five minutes.

[ 1317.333707] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=elementsd.service,mems_allowed=0,global_oom,task_memcg=/system.slice/elementsd.service,task=elementsd,pid=735,uid=998
[ 1317.333729] Out of memory: Killed process 735 (elementsd) total-vm:11785420kB, anon-rss:7676108kB, file-rss:1656kB, shmem-rss:0kB, UID:998 pgtables:19328kB oom_score_adj:0

**Comment by paarthh-dev** [comment]
I am facing the same issue on version: 23.3.3. I am already running bitcoin core on the same linux machine and now i started elements, it seemed fine initially, but then it started to consume more and more memory, i've also created swap file but that is also getting filled. on networking side the downloading rate is roughly 37Mbps when this starts happening, the machine has 2 cores and 2 Gb ram. Here is my elements.conf

```
chain=liquidv1
server=1
daemon=1
listen=0
prune=0
txindex=0

mainchainrpchost=127.0.0.1
mainchainrpcport=8332
mainchainrpccookiefile=cookie path
validatepegin=0

dbcache=100
maxconnections=20
#listenonion=0
#walletrbf=1
#fallbackfee=0.0001
#whitelist=127.0.0.1
maxmempool=100
par=2
```

this way it crashed due to OOM, then when i started the node next time it just filled the ram and swap and hanged without syncing blocks, i tried stopping it but after 10 minutes i have to force kill it, then i tried starting daemon again, same.
after deleting the blocks data it started syncing again.
Here is htop.

<img width="942" height="89" alt="Image" src="https://github.com/user-attachments/assets/5850e605-703f-4f4c-92e2-24da5ee62337" />

**Comment by jsarenik** [comment]
If not running as a daemon, what's the terminal (emulator) used? May that go OOM by buffering too many lines?

I run both bitcoind and elementsd with no issues in daemon mode.

Getinfo on each new block: https://anyone.eu.org/liquid/gi.txt

**Comment by delta1** [comment]
@paarthh-dev unfortunately 2GB memory is too little, you'll need at least 8GB and configure `trim_headers=1`, 16GB is preferred
